File: server/wordclimb/internal/domain/coins_test.go
package domain

import (
	"testing"
	"time"
)

// The vectors here are duplicated verbatim in apps/wordclimb/lib/coins.test.ts.
// They are the contract between the two implementations of the ISO week: if
// they drift, a player is refilled twice or not at all around the new year.
func Test_ISOWeekID(t *testing.T) {
	cases := []struct {
		date string
		want string
	}{
		// A plain midweek day.
		{"2026-08-30", "2026-W35"},
		// The year boundary, which is the whole reason this is not "the week
		// number since 1 January". These days belong to the next year's W01
		// because their Thursday does.
		{"2025-12-29", "2026-W01"},
		{"2025-12-31", "2026-W01"},
		{"2026-01-01", "2026-W01"},
		// ...and these belong to the previous year's last week.
		{"2027-01-01", "2026-W53"},
		// Monday starts a week, Sunday ends the same one.
		{"2026-08-24", "2026-W35"},
		{"2026-08-23", "2026-W34"},
		// Zero-padded, so the string sorts and compares as one shape.
		{"2026-01-05", "2026-W02"},
	}

	for _, c := range cases {
		at, err := time.Parse("2006-01-02", c.date)
		if err != nil {
			t.Fatalf("bad test date %q: %v", c.date, err)
		}
		if got := ISOWeekID(at); got != c.want {
			t.Errorf("ISOWeekID(%s) = %s, want %s", c.date, got, c.want)
		}
	}
}

func Test_ISOWeekID_isTimezoneIndependent(t *testing.T) {
	// Late evening in a timezone ahead of UTC is still the UTC day, or two
	// devices would disagree about when the week turns over.
	utc := time.Date(2026, 8, 30, 23, 0, 0, 0, time.UTC)
	ahead := utc.In(time.FixedZone("UTC+10", 10*60*60))

	if ISOWeekID(utc) != ISOWeekID(ahead) {
		t.Errorf("same instant gave %s and %s", ISOWeekID(utc), ISOWeekID(ahead))
	}
}

func Test_RefillCoins(t *testing.T) {
	now := time.Date(2026, 8, 30, 12, 0, 0, 0, time.UTC)
	thisWeek := ISOWeekID(now)

	t.Run("tops a spent balance up to the allowance", func(t *testing.T) {
		balance, week := RefillCoins(3, "2026-W34", now)
		if balance != WeeklyCoinAllowance || week != thisWeek {
			t.Errorf("got %d/%s, want %d/%s", balance, week, WeeklyCoinAllowance, thisWeek)
		}
	})

	t.Run("does not accumulate week over week", func(t *testing.T) {
		// Away for a month means ten coins waiting, not forty.
		if balance, _ := RefillCoins(0, "2026-W30", now); balance != WeeklyCoinAllowance {
			t.Errorf("got %d, want %d", balance, WeeklyCoinAllowance)
		}
	})

	t.Run("leaves a balance above the allowance alone", func(t *testing.T) {
		if balance, _ := RefillCoins(15, "2026-W34", now); balance != 15 {
			t.Errorf("got %d, want 15 — a hoarded balance must not be cut", balance)
		}
	})

	t.Run("does nothing twice in the same week", func(t *testing.T) {
		if balance, _ := RefillCoins(2, thisWeek, now); balance != 2 {
			t.Errorf("got %d, want 2 — refilled twice in one week", balance)
		}
	})

	t.Run("refills an account that has never been refilled", func(t *testing.T) {
		if balance, _ := RefillCoins(0, "", now); balance != WeeklyCoinAllowance {
			t.Errorf("got %d, want %d", balance, WeeklyCoinAllowance)
		}
	})
}

func Test_MaxCoinsForLadder(t *testing.T) {
	// A level's spend accumulates over every rung, so the ceiling has to scale
	// with the ladder. A flat one was worse than none: it clamped a real spend
	// down, the server answered with a balance higher than the player's, and
	// the client's reconcile handed the difference back as free coins.
	cases := []struct {
		rungs, want int
	}{
		{1, 5},
		{2, 10},
		{5, 25},
		{9, 45},
		// Defensive: an empty or malformed ladder still allows one rung rather
		// than zero, so a legitimate single hint is not clawed back.
		{0, 5},
		{-3, 5},
	}

	for _, c := range cases {
		if got := MaxCoinsForLadder(c.rungs); got != c.want {
			t.Errorf("MaxCoinsForLadder(%d) = %d, want %d", c.rungs, got, c.want)
		}
	}
}

func Test_DebitCoins(t *testing.T) {
	cases := []struct {
		name                  string
		balance, spent, limit int
		want                  int
	}{
		{"ordinary spend", 10, 3, 25, 7},
		{"spending nothing", 10, 0, 25, 10},
		{"exactly empties", 3, 3, 25, 0},
		// A whole ladder solved on hints is a legitimate spend, and must be
		// charged in full rather than clamped — clamping refunded the rest.
		{"charges a full multi-rung spend", 30, 27, 45, 3},
		// Offline spending is only reported when the level is banked, by which
		// time the server may have reconciled: the player must not end up in
		// debt through a race they cannot see.
		{"never goes negative", 2, 3, 25, 0},
		// A tampered payload must not be able to zero an account...
		{"clamps a claim beyond what the ladder allows", 50, 999, 10, 40},
		// ...nor mint coins by spending a negative amount.
		{"refuses a negative spend", 10, -5, 25, 10},
		{"treats a negative limit as zero", 10, 4, -1, 10},
	}

	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			if got := DebitCoins(c.balance, c.spent, c.limit); got != c.want {
				t.Errorf("DebitCoins(%d, %d, %d) = %d, want %d",
					c.balance, c.spent, c.limit, got, c.want)
			}
		})
	}
}

File: apps/wordclimb/lib/coins.test.ts
import { beforeEach, describe, expect, it } from "vitest"
import {
  HINT_COSTS,
  WEEKLY_COIN_ALLOWANCE,
  canAfford,
  getCoinBalance,
  isoWeekId,
  reconcileCoins,
  spendCoins,
} from "./coins"

const at = (iso: string) => new Date(`${iso}T12:00:00Z`)

function storedCoins() {
  const raw = window.localStorage.getItem("wordclimb-coins")
  return raw ? JSON.parse(raw) : null
}

describe("isoWeekId", () => {
  // These vectors are duplicated verbatim in
  // server/wordclimb/internal/domain/coins_test.go. They are the contract
  // between the two implementations: if they drift, a player is refilled twice
  // or not at all around the new year, where "week 1" and "the first week of
  // January" are not the same thing.
  it.each([
    ["2026-08-30", "2026-W35"],
    ["2025-12-29", "2026-W01"],
    ["2025-12-31", "2026-W01"],
    ["2026-01-01", "2026-W01"],
    ["2027-01-01", "2026-W53"],
    ["2026-08-24", "2026-W35"],
    ["2026-08-23", "2026-W34"],
    ["2026-01-05", "2026-W02"],
  ])("%s falls in %s", (date, expected) => {
    expect(isoWeekId(at(date))).toBe(expected)
  })

  it("does not shift with the device's timezone", () => {
    // Late evening UTC is a different local date east of Greenwich; both must
    // still report the same week, or two devices disagree on when it turns.
    const utc = new Date("2026-08-30T23:00:00Z")
    const sameInstantElsewhere = new Date(utc.getTime())
    expect(isoWeekId(utc)).toBe(isoWeekId(sameInstantElsewhere))
  })
})

describe("coin balance", () => {
  beforeEach(() => {
    window.localStorage.clear()
  })

  it("starts a new player off with a full week's allowance", () => {
    expect(getCoinBalance(at("2026-08-30"))).toBe(WEEKLY_COIN_ALLOWANCE)
  })

  it("tops up when a new week starts", () => {
    spendCoins(7, at("2026-08-30"))
    expect(getCoinBalance(at("2026-08-30"))).toBe(3)

    expect(getCoinBalance(at("2026-09-01"))).toBe(WEEKLY_COIN_ALLOWANCE)
  })

  it("does not accumulate week over week", () => {
    // Away for a month means ten coins waiting, not forty.
    spendCoins(10, at("2026-07-06"))
    expect(getCoinBalance(at("2026-08-30"))).toBe(WEEKLY_COIN_ALLOWANCE)
  })

  it("leaves a balance above the allowance alone", () => {
    reconcileCoins(50, at("2026-08-30"))
    expect(getCoinBalance(at("2026-09-01"))).toBe(50)
  })

  it("refills once, not on every read", () => {
    getCoinBalance(at("2026-09-01"))
    spendCoins(4, at("2026-09-01"))
    expect(getCoinBalance(at("2026-09-01"))).toBe(6)
  })
})

describe("spending", () => {
  beforeEach(() => {
    window.localStorage.clear()
  })

  it("charges the hint's price", () => {
    expect(spendCoins(HINT_COSTS.fullWord, at("2026-08-30"))).toBe(true)
    expect(getCoinBalance(at("2026-08-30"))).toBe(
      WEEKLY_COIN_ALLOWANCE - HINT_COSTS.fullWord
    )
  })

  it("refuses rather than going into debt", () => {
    spendCoins(9, at("2026-08-30"))
    expect(canAfford(HINT_COSTS.fullWord, at("2026-08-30"))).toBe(false)
    expect(spendCoins(HINT_COSTS.fullWord, at("2026-08-30"))).toBe(false)
    // The refused spend must not have moved anything.
    expect(getCoinBalance(at("2026-08-30"))).toBe(1)
  })

  it("prices revealing the whole word above the informative hints", () => {
    // It does not help the player find the word, it hands it over.
    expect(HINT_COSTS.fullWord).toBeGreaterThan(HINT_COSTS.firstLetter)
    expect(HINT_COSTS.fullWord).toBeGreaterThan(HINT_COSTS.definition)
  })
})

describe("reconciling with the server", () => {
  beforeEach(() => {
    window.localStorage.clear()
  })

  it("takes the server's number as the truth, in both directions", () => {
    reconcileCoins(4, at("2026-08-30"))
    expect(getCoinBalance(at("2026-08-30"))).toBe(4)

    reconcileCoins(25, at("2026-08-30"))
    expect(getCoinBalance(at("2026-08-30"))).toBe(25)
  })

  it("never stores a negative balance", () => {
    reconcileCoins(-5, at("2026-08-30"))
    expect(getCoinBalance(at("2026-08-30"))).toBe(0)
  })

  it("survives corrupted storage rather than throwing", () => {
    window.localStorage.setItem("wordclimb-coins", "not json")
    expect(getCoinBalance(at("2026-08-30"))).toBe(WEEKLY_COIN_ALLOWANCE)
    expect(storedCoins()).toEqual({
      balance: WEEKLY_COIN_ALLOWANCE,
      week: "2026-W35",
    })
  })
})
